core/libp2p/network/negotiator.hpp

```diff
@@ -0,0 +1,58 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef KAGOME_NEGOTIATOR_HPP
+#define KAGOME_NEGOTIATOR_HPP
+
+#include <memory>
+
+#include "libp2p/basic/readwritecloser.hpp"
+#include "libp2p/peer/protocol.hpp"
+#include "libp2p/connection/stream.hpp"
+
+namespace libp2p::network {
+
+  /**
+   * @brief Negotiator is a component capable of reaching agreement over what
+   * protocols to use for inbound streams of communication.
+   * @see
+   * https://sourcegraph.com/github.com/libp2p/go-libp2p-core@consolidate-skeleton/-/blob/protocol/switch.go#L53
+   */
+  struct Negotiator {
+    virtual ~Negotiator() = default;
+
+    // TODO: come up with better name
+    struct Return {
```

Comment:
Negotiation result can be called Agreement, but I'm not sure that result is good. Protocol is result of negitiations. How do we obtain handler?

Reply:
this class is a subject for change